app/scripts/modules/core/securityGroup/securityGroupReader.service.ts

```diff
@@ -0,0 +1,412 @@
+import {filter, forOwn, get, uniq} from 'lodash';
+import {module} from 'angular';
+
+import {API_SERVICE, Api} from 'core/api/api.service';
+import {NAMING_SERVICE, NamingService, IComponentName} from 'core/naming/naming.service';
+import {INFRASTRUCTURE_CACHE_SERVICE, InfrastructureCacheService} from 'core/cache/infrastructureCaches.service';
+import {Application} from 'core/application/application.model';
+import {ISecurityGroup, ILoadBalancer, ServerGroup} from 'core/domain';
+import {SECURITY_GROUP_TRANSFORMER_SERVICE, SecurityGroupTransformerService} from './securityGroupTransformer.service';
+
+interface IRegionAccount {
+  account: string;
+  accountName: string;
+  id: string;
+  name: string;
+  provider: string;
+  region: string;
+  vpcId: string;
+}
+
+interface IRegions {
+  [key: string]: IRegionAccount[]; // regionName
+}
+
+interface IProviders {
+  [key: string]: IRegions; // providerName
+}
+
+interface IGroupsByRegion {
+  [key: string]: IRegionAccount[]; // regionName
+}
+
+interface IGroupsByProvider {
+  [key: string]: IGroupsByRegion; // providerName
+}
+
+interface ISecurityGroupsByRegion {
+  [key: string]: IRegionAccount;  // regionName
+}
+
+interface ISecurityGroupsByAccount {
+  [key: string]: ISecurityGroupsByRegion;  // accountName
+}
+
+interface IIndexedSecurityGroups {
+  [key: string]: ISecurityGroupsByAccount;
+}
+
+export interface IApplicationSecurityGroup {
+  name: string;
+}
+
+interface IReaderSecurityGroup extends ISecurityGroup {
+  securityGroups: IGroupsByRegion;
+}
+
+interface ISecurityGroupDetailInboundRule {
+  accountName: string;
+  id: string;
+  inferredName: boolean;
+  name: string;
+}
+
+interface ISecurityGroupDetailRule {
+  range: any;
+  securityGroup: ISecurityGroupDetailInboundRule;
+}
+
+interface ISecurityGroupProcessorResult {
+  notFoundCaught: boolean;
+  securityGroups: ISecurityGroup[];
+}
+
+export interface IGroupsByAccount {
+  [key: string]: IProviders; // accountName
+}
+
+export interface ISecurityGroupDetail {
+  inboundRules: ISecurityGroupDetailRule[];
+  ipRangeRules: ISecurityGroupDetailRule[];
```

Comment:
This will mostly work, but it doesn't map completely cleanly.  While `IpRangeRule` has a `range` property, `SecurityGroupRule` does not.  The legacy code implies that a "hybrid" version is sent to the UI:

```javascript
details.ipRangeRules = details.inboundRules.filter(function(rule) {
  return rule.range;
});
details.securityGroupRules = details.inboundRules.filter(function(rule) {
  return rule.securityGroup;
});
```

I'll need to take this into consideration for fixing the interface types.